coderd/jobreaper/detector.go

```diff
@@ -172,34 +189,42 @@ func (d *Detector) run(t time.Time) Stats {
 		Error:            nil,
 	}
 
-	// Find all provisioner jobs that are currently running but have not
-	// received an update in the last 5 minutes.
-	jobs, err := d.db.GetHungProvisionerJobs(ctx, t.Add(-HungJobDuration))
+	// Find all provisioner jobs to be reaped
+	jobs, err := d.db.GetProvisionerJobsToBeReaped(ctx, database.GetProvisionerJobsToBeReapedParams{
+		PendingSince: t.Add(-PendingJobDuration),
+		HungSince:    t.Add(-HungJobDuration),
+		MaxJobs:      MaxJobsPerRun,
+	})
 	if err != nil {
-		stats.Error = xerrors.Errorf("get hung provisioner jobs: %w", err)
+		stats.Error = xerrors.Errorf("get provisioner jobs to be reaped: %w", err)
 		return stats
 	}
 
-	// Limit the number of jobs we'll unhang in a single run to avoid
-	// timing out.
-	if len(jobs) > MaxJobsPerRun {
-		// Pick a random subset of the jobs to unhang.
-		rand.Shuffle(len(jobs), func(i, j int) {
-			jobs[i], jobs[j] = jobs[j], jobs[i]
-		})
-		jobs = jobs[:MaxJobsPerRun]
-	}
+	jobsToReap := make([]*jobToReap, 0, len(jobs))
 
-	// Send a message into the build log for each hung job saying that it
-	// has been detected and will be terminated, then mark the job as
-	// failed.
 	for _, job := range jobs {
+		j := &jobToReap{
+			ID: job.ID,
+		}
+		if job.JobStatus == database.ProvisionerJobStatusPending {
+			j.Threshold = PendingJobDuration
+			j.Type = Pending
+		} else {
+			j.Threshold = HungJobDuration
+			j.Type = Hung
+		}
+		jobsToReap = append(jobsToReap, j)
+	}
+
+	// Send a message into the build log for each hung or pending job saying that it
+	// has been detected and will be terminated, then mark the job as failed.
+	for _, job := range jobsToReap {
 		log := d.log.With(slog.F("job_id", job.ID))
 
-		err := unhangJob(ctx, log, d.db, d.pubsub, job.ID)
+		err := reapJob(ctx, log, d.db, d.pubsub, job)
 		if err != nil {
 			if !(xerrors.As(err, &acquireLockError{}) || xerrors.As(err, &jobIneligibleError{})) {
-				log.Error(ctx, "error forcefully terminating hung provisioner job", slog.Error(err))
+				log.Error(ctx, fmt.Sprintf("error forcefully terminating %s provisioner job", job.Type), slog.Error(err))
```

Comment:
I'd just change this to be `"error forcefully terminating provisioner job", slog.F("type", job.Type), slog.Error(err)`